internal/factsengine/gatherers/dir_scan.go

```diff
@@ -0,0 +1,194 @@
+package gatherers
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"syscall"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
+
+	"github.com/trento-project/agent/pkg/factsengine/entities"
+)
+
+const (
+	DirScanGathererName = "dir_scan"
+)
+
+// nolint:gochecknoglobals
+var (
+	DirScanMissingArgumentError = entities.FactGatheringError{
+		Type:    "dir-scan-missing-argument",
+		Message: "missing required argument",
+	}
+
+	DirScanOpenError = entities.FactGatheringError{
+		Type:    "dir-scan-open-error",
+		Message: "could not open the provided directory",
+	}
+
+	DirScanScanningError = entities.FactGatheringError{
+		Type:    "dir-scan-scanning-error",
+		Message: "error during directory scanning",
+	}
+)
+
+type DirScanDetails struct {
+	Name  string           `json:"name,omitempty"`
+	Owner string           `json:"owner"`
+	Group string           `json:"group"`
+	Files []DirScanDetails `json:"files,omitempty"`
+}
+
+type DirScanStatInfo struct {
+	Group string
+	Owner string
+}
+
+type DirScanResult map[string]DirScanDetails
+
+//go:generate mockery --name=UserSearcher
+type UserSearcher interface {
+	GetUsernameByID(userID string) (string, error)
+}
+
+//go:generate mockery --name=GroupSearcher
+type GroupSearcher interface {
+	GetGroupByID(groupID string) (string, error)
+}
+
+type DirScanGatherer struct {
+	fs            afero.Fs
+	userSearcher  UserSearcher
+	groupSearcher GroupSearcher
+}
+
+func NewDirScanGatherer(fs afero.Fs, userSearcher UserSearcher, groupSearcher GroupSearcher) *DirScanGatherer {
+	return &DirScanGatherer{fs: fs, userSearcher: userSearcher, groupSearcher: groupSearcher}
+}
+
+func NewDefaultDirScanGatherer() *DirScanGatherer {
+	cf := CredentialsFetcher{}
+	return NewDirScanGatherer(afero.NewOsFs(), &cf, &cf)
+}
+
+func (d *DirScanGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	log.Infof("Starting %s facts gathering process", DirScanGathererName)
+	facts := []entities.Fact{}
+
+	for _, requestedFact := range factsRequests {
+		if requestedFact.Argument == "" {
+			facts = append(facts, entities.NewFactGatheredWithError(requestedFact, &DirScanMissingArgumentError))
```

Comment:
I'd add a `log.Error(DirScanMissingArgumentError.Message)` here, might be helpful when diagnosing issues

Reply:
done